default/python/AI/ResourcesAI.py

```diff
@@ -78,47 +126,46 @@ def __init__(self):
         universe = fo.getUniverse()
 
         resource_timer.start("getPlanets")
-        planet_ids = list(PlanetUtilsAI.get_owned_planets_by_empire())
+        self.planet_ids = [planet.id for planet in PlanetUtilsAI.get_empire_populated_planets()]
 
         resource_timer.start("Targets")
-        self.all_planet_info = {pid: PlanetFocusInfo(universe.getPlanet(pid)) for pid in planet_ids}
-
-        self.raw_planet_info = dict(self.all_planet_info)
+        self.planet_info = {pid: PlanetFocusInfo(universe.getPlanet(pid)) for pid in self.planet_ids}
         self.baked_planet_info = {}
-        for pid, pinfo in list(self.raw_planet_info.items()):
-            if not pinfo.planet.availableFoci:
-                self.baked_planet_info[pid] = self.raw_planet_info.pop(pid)
         aistate = get_aistate()
-        self.priority = (
-            aistate.get_priority(PriorityType.RESOURCE_PRODUCTION),
-            aistate.get_priority(PriorityType.RESOURCE_RESEARCH),
-            aistate.get_priority(PriorityType.RESOURCE_INFLUENCE),
-        )
+        self.priority_industry = aistate.get_priority(PriorityType.RESOURCE_PRODUCTION)
+        self.priority_research = aistate.get_priority(PriorityType.RESOURCE_RESEARCH)
+        self.priority_influence = aistate.get_priority(PriorityType.RESOURCE_INFLUENCE)
         self.have_bureaucracy = bureaucracy in fo.getEmpire().adoptedPolicies
-
-    def bake_future_focus(self, pid, focus, update=True, force=False):
+        self.have_liberty = liberty in fo.getEmpire().adoptedPolicies
+        self.have_automation = tech_is_complete(AIDependencies.PRO_AUTO_1)
+        self.have_ai = tech_is_complete(AIDependencies.LRN_ARTIF_MINDS_1)
```

Comment:
I do not see the advantage of using a function over a member variable.
Memory? This class is created only once per turn.
Plus, surely a boolean member is faster than a string lookup.

Reply:
Memory and performance are not the cases.  

It's more about coupling.   So methods that use that values get them indirectly, via class attributes.  Instead of giving the information to the method, you are giving it to the class, and the class gove it to the method. 

There are some valid cases when we need class attributes, value is present only at creation time/place, and you need to preserve the state of a mutable object, cashing for performance.  In other cases, I prefer to keep it as simple as possible.